### ISSUE TITLE
Define package types for ASF projects (Apache Software Foundation)

### ISSUE THREAD

**Issue by pombredanne** [package-url/purl-spec#305]
We should define a package type for ASF projects (Apache Software Foundation)

The spec mentioned originally `apache for Apache projects packages`. The direction may be to use `asf` rather than `apache`.

There have been on-going discussion on the ASF mailing lists on the topic and we need to collect these links for reference and invite the ASF folks to join and help define this (important) package type!

@raboof ping

**Comment by brianf** [comment]
Do you have the links to the threads? I'm curious what the use case is. Project !=Package. In fact I have asserted for >10 years that the major problem with CPE is that it maps to just a project... where a project like Struts has ~80 packages, making it useless for most use cases. Having a pURL recreate that lossy coordinate would be a huge step backwards.

**Comment by raboof** [comment]
> Do you have the links to the threads?

https://lists.apache.org/thread/vc3h1t7plq3sgtqvp385s4nlo3l7rry7 and https://lists.apache.org/thread/75l9f8bcs9fm232p2j3prbj9fw2or2k5 come to mind.

> the major problem with CPE is that it maps to just a project... where a project like Struts has ~80 packages, making it useless for most use cases. Having a pURL recreate that lossy coordinate would be a huge step backwards.

That would be good to flesh out. I could see an approach where we use the PMC id as the first segment, and the PMC can determine whether/how to add further detail - something like `pkg:asf/celix` could perhaps stand on its own, while struts might introduce `pkg:asf/struts/oval-plugin` etc for its various components. We should probably give some guidance on how to apply that. WDYT?

**Comment by stevespringett** [comment]
Most Apache projects fall into existing support for package ecosystems already supported by purl. See https://projects.apache.org/projects.html?language

Per definition, a purl is:
> a URL string used to identify and locate a software package...

I cannot locate `pkg:asf/struts/oval-plugin`. Is on Maven Central or somewhere else? Additionally, `oval-plugin` already has a purl which is:

`pkg:maven/org.apache.struts/struts2-oval-plugin@x.x.x` therefore adding `pkg:asf/struts/oval-plugin` would introduce confusion IMO.

**Comment by ppkarwasz** [comment]
> That would be good to flesh out. I could see an approach where we use the PMC id as the first segment, and the PMC can determine whether/how to add further detail - something like `pkg:asf/celix` could perhaps stand on its own, while struts might introduce `pkg:asf/struts/oval-plugin` etc for its various components. We should probably give some guidance on how to apply that. WDYT?

Yes, we could follow the directory structure in `downloads.apache.org` (or `archive.apache.org`) to provide the coordinates of a **source** archive (which is the only official product of the ASF). For example:

- `pkg:apache/logging/log4j@2.24.3`,
- `pkg:apache/struts@7.0.0`,
- `pkg:apache/tomcat/tomcat-11%2Fsrc@v11.0.4`,
- `pkg:apache/commons/lang%2Fsource@3.17.0`.

However, as the examples show, the location of the source archives varies greatly between PMCs and even varies in time (older Log4j Kotlin API where in `logging/log4j/kotlin`, but newer are in `logging/log4j-kotlin`). The naming of the source archives is even less regular.

I think that right now there is no sense in introducing an `asf` type. We could think about that when Apache Trusted Releases is out and forces some order in this chaos.

**Comment by stevespringett** [comment]
Since this ticket was originally created, it is now possible to use the SWID PURL type.

For example:
```
pkg:swid/Apache Software Foundation/apache.org/logging/log4@2.24.3?tag_id=de20a672-4e41-449a-adf7-945f978ede9d
```

**Comment by stevespringett** [comment]
This is a big change. If PURL is going to support both ecosystems and vendors for the type, then I would be prepared for an influx of organizations and foundations trying to reserve names. At least with SWID, they would be able to use it without having an influx of PURL types being defined. Additionally, what happens when two vendors, both named Acme, try to use it? Using `apache` without using DNS, one might assume Apache Corporation instead of Apache Software Foundation.

The other big change is location. In the case of Apache, its fairly simple, where they have a `/logging/log4j` directory structure, but what happens when precedence has been set that vendors now can have PURL types and Acme's download server uses something like `https://download.example.com?fid=0123456789`. I certainly would not want to end up with something like `pkg:/acme/0123456789` or (worse) `pkg:/acme/my%20application?fid=0123456789` where `0123456789` is the unique identifier to download the package.

I would seriously consider using either CPE or the SWID PURL type instead of introducing vendor-specific PURL types. If the SWID PURL type needs revisiting, thats fine, we can do that. But I have a lot of reservations about the precedence this would set, not about ASF specifically.

**Comment by ppkarwasz** [comment]
I agree that vendor-specific types are a dangerous precedent, but:

- CPE is not an URL, in the sense that it does not allow to locate a product.
- SWID is behind a paying wall: does it guarantee uniqueness of the generated identifiers?

Some DNS-based type would be probably ideal with a syntax like `pkg:dns/<domain>/<domain_dependent_path>/<product>`. The "default" download URL could be provided in a metadata file in a `.well-known` of `<domain>`.

**Comment by raboof** [comment]
> I agree that vendor-specific types are a dangerous precedent

That has been part of my hesitation as well: "we" (generally) need a way to refer to software that's not in any of the defined ecosystems. Assuming we have a good way to do that, perhaps we don't need a separate namespace for "ASF software that is not in any of the defined namespaces". OTOH, especially with the '

>     * CPE is not an URL, in the sense that it does not allow to locate a product.

I'm not sure if that's really a requirement? I would be OK with just being able to 'refer to' software tbh, and leave the 'location' of products to be defined elsewhere.

The problem I see with CPE is that it's not federated, so you need to talk to the NVD CPE team (https://nvd.nist.gov/products/cpe) *before* using a new CPE. This would mean an explosion of the CPE dictionary if you'd want to use it to refer to software in e.g. an SBOM - I just don't see how that would ever scale in practice.

Also @brianf mentioned above that you cannot use CPE to refer to 'sub-artifacts' of projects, which would be a problem.

>     * SWID is behind a paying wall: does it guarantee uniqueness of the generated identifiers?

I think https://datatracker.ietf.org/doc/rfc9393/ lets you 'read between the lines' of what 'regular' SWID looks like as well. Here the importance of the mandatory globally unique 'tagId' makes it awkward to use in our context. I'm sure we could find a 'hack' to make it 'technically work', but...

> Some DNS-based type would be probably ideal with a syntax like `pkg:dns/<domain>/<domain_dependent_path>/<product>`. The "default" download URL could be provided in a metadata file in a `.well-known` of `<domain>`.

something like this, or even the `generic` type (perhaps with some conventions around it) might then be a better fit?

OTOH, perhaps we have enough 'structure' in the ASF to justify having an `asf` Purl type to mirror that structure - I do think the examples @pombredanne gives above are compelling...

**Comment by stevespringett** [comment]
For SWID specifically, refer to https://package-url.github.io/purl-swid-generator-ui/ for an easy way to generate them - and it relies on DNS per the specification. While the SWID spec doesn't strictly enforce UUIDs, is strongly recommends them, so yes, in most cases the SWID identifier will be globally unique.

**Comment by raboof** [comment]
My impression is we should close this issue in favor of the more generic solutions in development in https://github.com/package-url/purl-spec/issues/784 / https://github.com/package-url/purl-spec/issues/516 - agreed?
